File: leetcode/c++/001.cpp
#include <vector>
using std::vector;
#include <iostream>
#include <unordered_map>
using std::unordered_map;
using namespace std;

class Solution {
public:
  vector<int> twoSum(vector<int>& nums, int target)   
  {
    unordered_map<int, int> d;
    vector<int> ret;
    for (size_t i = 0; i < nums.size(); ++i)
    {
      if (d.find(nums[i]) == d.end())
      {
        d[target-nums[i]] = i;
      }
      else
      {
        ret.push_back(d[nums[i]]);
        ret.push_back(i);
        break;
      }
    }
    return ret;
  }
};

int main() {
  Solution s;
  int arr[5] = {1, 2, 3, 4, 5};
  vector<int> nums(arr, arr + 5);
  vector<int> res = s.twoSum(nums, 6);
  for (int i = 0; i < res.size(); i++)
    cout<<res.at(i)<<endl;
  return 0;
}

File: leetcode/001.js
/*
给定一个整数数组和一个目标值，找出数组中和为目标值的两个数。

你可以假设每个输入只对应一种答案，且同样的元素不能被重复利用。
z
示例:

给定 nums = [2, 7, 11, 15], target = 9

因为 nums[0] + nums[1] = 2 + 7 = 9
所以返回 [0, 1]
*/

var twoSum = function(nums, target) {
  var hash = {}
  for (let i = 0; i < nums.length; i++) {
    const num = nums[i]
    if (hash[num] !== undefined) {
      return [hash[num], i]
    } else {
      hash[target - num] = i
    }
  }
  return []
}
